File: cpp/29_remove_nth_node.cpp
/**
 * LeetCode #19: 删除链表的倒数第 N 个节点
 * 题目: 删除倒数第 n 个节点，返回头节点。
 * 解法: 前后指针（距离差）- 快指针先走 n+1 步，然后快慢同速前进。
 *       当快指针到达末尾时，慢指针的下一个就是待删除节点。
 * 时间 O(n)，空间 O(1)。
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x=0, ListNode* n=nullptr) : val(x), next(n) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy(0, head);
        ListNode *fast = &dummy, *slow = &dummy;
        for (int i = 0; i < n + 1; ++i) fast = fast->next;
        while (fast) {
            fast = fast->next;
            slow = slow->next;
        }
        slow->next = slow->next->next;
        return dummy.next;
    }
};

int main() {
    Solution sol;
    ListNode* head = new ListNode(1,
                    new ListNode(2,
                    new ListNode(3,
                    new ListNode(4,
                    new ListNode(5)))));
    ListNode* res = sol.removeNthFromEnd(head, 2);
    while (res) { cout << res->val << " "; res = res->next; } // 1 2 3 5
    cout << endl;
    return 0;
}

File: python/29_remove_nth_node.py
#!/usr/bin/env python3
"""
LeetCode #19: 删除链表的倒数第 N 个节点
题目: 给定链表和整数 n，删除倒数第 n 个节点，返回头节点。
解法: 前后指针（距离差） - 让快指针先走 n+1 步，然后快慢同速前进，
      当快指针到达尾部时，慢指针的下一个就是待删除节点。
      时间 O(n)，空间 O(1)。
"""


class ListNode:
    def __init__(self, x=0, next=None):
        self.val = x
        self.next = next


def removeNthFromEnd(head: ListNode, n: int) -> ListNode:
    """
    前后指针: fast 先走 n+1 步，然后 fast 和 slow 同时前进。
    当 fast 到达末尾时，slow.next 即待删除节点。
    """
    dummy = ListNode(0, head)
    fast = slow = dummy
    for _ in range(n + 1):
        fast = fast.next
    while fast:
        fast = fast.next
        slow = slow.next
    slow.next = slow.next.next
    return dummy.next


if __name__ == "__main__":
    # 1->2->3->4->5, n=2 → 1->2->3->5
    nodes = [ListNode(i) for i in range(1, 6)]
    for i in range(4): nodes[i].next = nodes[i+1]
    res = removeNthFromEnd(nodes[0], 2)
    vals = []
    while res:
        vals.append(res.val)
        res = res.next
    print(vals)  # [1, 2, 3, 5]
